### ISSUE TITLE
Conch integration tests fail because DSA is deprecated in OpenSSH 7.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8180]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| judge_gregg reported|
|-|-|
|Trac ID|trac#8180|
|Type|defect|
|Created|2016-01-18 19:52:26Z|
|Branch|https://github.com/twisted/twisted/tree/conch-integration-fail-openssh7-8180|

Tests affected are:

twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_exec
twisted.conch.test.test_conch.OpenSSHClientRekeyTests.test_clientRekey
twisted.conch.test.test_conch.OpenSSHKeyExchangeTestCase.test_DH_GROUP1
twisted.conch.test.test_conch.OpenSSHKeyExchangeTestCase.test_DH_GROUP14
twisted.conch.test.test_conch.OpenSSHKeyExchangeTestCase.test_DH_GROUP_EXCHANGE_SHA1
twisted.conch.test.test_conch.OpenSSHKeyExchangeTestCase.test_DH_GROUP_EXCHANGE_SHA256
twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwarding
twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_localToRemoteForwardingV6
twisted.conch.test.test_conch.OpenSSHClientForwardingTests.test_remoteToLocalForwarding

Fix is fairly trivial: add PubkeyAcceptedKeyTypes=ssh-dss to command line options.

Attachments:

* [conch-integration-fail-openssh7-8180-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f6c/f6c62b69aead29b5b86258d17e6a52b006958954/e04acf8a484ffe0a13b4cb6a74fcb8b9e3971c0b.patch) (1316 bytes) - added by judge_gregg on 2016-01-18 19:57:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8180 8180
type__defect defect
reporter__judge_gregg judge_gregg
priority__low low
milestone__None None
branch__branches_conch_integration_fail_openssh7_8180 branches/conch-integration-fail-openssh7-8180
branch_author__mithrandi__judge_gregg mithrandi, judge_gregg
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1453146746705519 1453146746705519
changetime__1454301919054184 1454301919054184
version__None None
owner__mithrandi mithrandi
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(In [46640]) Branching to allow-dsa-for-now-8180.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @glyph|
|-|-|

Thanks for noticing this, and for submitting a patch.  It would be nice to get out in front of one of these integration issues for once, rather than upgrading my OS and discovering that I can no longer run the Twisted tests :-). I will probably commit if none of our current builders complain, since it's just a change to some test data.  However, separately, we should probably be looking at deprecating DSA ourselves if OpenSSH has done it...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to judge_gregg|
|-|-|

It looks like although it would work with OpenSSH 7, this code [fails badly on our existing support matrix](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/allow-dsa-for-now-8180).  Would you mind writing some code that detects whether the option is supported / necessary and add it only in that case?

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_6"></a>|@mithrandi commented|
|-|-|

(In [46677]) Branching to conch-integration-fail-openssh7-8180.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_7"></a>|@mithrandi removed owner|
|-|-|

Buildbot is green now: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-integration-fail-openssh7-8180

I added `-F /dev/null` to the command lines because local configuration (such as `ControlMaster`, or other crypto settings) can cause spurious problems or failures.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @mithrandi|
|-|-|

Thanks for the patch.

For reference: OpenSSH 7.0 release announcement http://lists.mindrot.org/pipermail/openssh-unix-announce/2015-August/000122.html


I think it needs an in-line comment to describe why 'ssh -V' is called first... and it looks like duplicate code.

Maybe we can have a helper called `getOpenSSHDefaultArguments()` which will handle adding ```-F /dev/null``` and ```PubkeyAcceptedKeyTypes``` options.

I think that the code needs an in-line comment to describe why `-F /dev/null` is needed... as I assume that many of the people reading the code will not search for this ticket and check each comment.

There was a spurious failure on freebsd10.1-py2.7 related to counch. I manually trigger a new test run and it is now green.

I have no idea how to be pro-active about these incompatibility changes... but I think there is no way as we can not 
predict when and how OpenSSH will change.

Only minor comments.

Please check my comments and merge at will.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_9"></a>|@mithrandi commented|
|-|-|

* I added some comments explaining the use of `ssh -V` and `-F /dev/null`.
* I considered the helper function idea, but the use of `PubkeyAcceptedKeyTypes ssh-dss` is only applicable to those tests which use a DSA host key; so `-F /dev/null` is the only option we definitely want anywhere, which probably doesn't warrant a helper function on its own, at least not yet.
* As far as being proactive goes, a "bleeding edge" builder may help with this (as per discussion in [#8187](https://github.com/twisted/twisted/issues/8187) and now on the mailing list, hopefully)

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_10"></a>|@mithrandi set status to `closed`|
|-|-|

(In [46701]) Merge conch-integration-fail-openssh7-8180: Make conch integration tests work with OpenSSH 7.0+

Author: judge_gregg, mithrandi
Reviewer: adiroiban
Fixes: [#8180](https://github.com/twisted/twisted/issues/8180)

OpenSSH 7.0 disables DSA (ssh-dss) host and client keys by default, so we need
to do some extra work to reenable these in the integration tests that use DSA
host keys.
